### PR TITLE
feat/expose state and key as properties in DMEnvWrapper

### PR DESCRIPTION
**Pull request by louis-csm**
add state and key as properties in DMEnvWrapper class https://github.com/instadeepai/jumanji/issues/27

### CODE REVIEW

File: jumanji/wrappers.py [hunk 1]
```diff
@@ -215,6 +215,22 @@ def action_spec(self) -> dm_env.specs.Array:
...
+    def state(self, new_state: Any) -> None:
+        self._state = new_state
+
+    @property
+    def key(self) -> PRNGKey:
+        return self._key
+
+    @key.setter
+    def key(self, new_key: PRNGKey) -> None:
+        self._key = new_key
```
[the remark this answers is not in the document]
Comment (louis-csm) [reply jumanji/wrappers.py:232]: It is not what has been asked in the issue https://github.com/instadeepai/jumanji/issues/27

[the remark this answers is not in the document]
Comment (louis-csm) [reply jumanji/wrappers.py:232]: I can but it is not what has been asked in [the issue](https://github.com/instadeepai/jumanji/issues/27). Sorry I forgot to link the pr to the issue

[the remark this answers is not in the document]
Comment (cemlyn007) [reply jumanji/wrappers.py:232]: Sorry, I have discussed this with @clement-bonnet and we agreed to expose these as public attributes because that is a more compact/cleaner solution.

Thank you for your interest and I am looking forward to reviewing your final contribution. 🙂

**Review by cemlyn007 — CHANGES_REQUESTED** [review]
Requesting that the attributes are made public instead of creating getter and setter properties.

File: jumanji/wrappers.py [hunk 2]
```diff
@@ -215,6 +215,22 @@ def action_spec(self) -> dm_env.specs.Array:
     def unwrapped(self) -> Environment:
         return self._env
 
+    @property
+    def state(self) -> Any:
```
Comment (clement-bonnet) [inline jumanji/wrappers.py:219]: Could we use `State` instead of `Any` here?
```suggestion
    def state(self) -> State:
```
Would you mind applying the same to `self._state` in the `__init__`?

File: jumanji/wrappers.py [hunk 3]
```diff
@@ -215,6 +215,22 @@ def action_spec(self) -> dm_env.specs.Array:
...
+        return self._state
+
+    @state.setter
+    def state(self, new_state: Any) -> None:
```
Comment (clement-bonnet) [inline jumanji/wrappers.py:223]: Same here

**Comment by clement-bonnet** [comment]
Could we please implement a test in the `TestJumanjiEnvironmentToDeepMindEnv` class of `wrappers_test.py` for getting and setting the key and state attributes?

**Review by clement-bonnet — CHANGES_REQUESTED** [review]
Nice work!

File: jumanji/wrappers_test.py [hunk 4]
```diff
@@ -228,6 +235,25 @@ def test_jumanji_environment_to_deep_mind_env__unwrapped(
         """Validates unwrapped property of the wrapped environment."""
         assert isinstance(fake_dm_env.unwrapped, Environment)
 
+    def test_jumanji_environment_to_deep_mind_env__key_getter_setter(
+        self, fake_environment: FakeEnvironment)-> None:
+        """validate state getter and setter"""
```
Comment (clement-bonnet) [inline jumanji/wrappers_test.py:240]: NIT: Could we please update docstrings to have them consistent with others? We try to keep them case sensitive.
```suggestion
        """Validates state getter and setter."""
```

File: jumanji/wrappers_test.py [hunk 5]
```diff
@@ -228,6 +235,25 @@ def test_jumanji_environment_to_deep_mind_env__unwrapped(
...
+        fake_dm_env = JumanjiToDMEnvWrapper(fake_environment, key=key)
+        assert jnp.array_equal(key, fake_dm_env.key)
+        new_key = jax.random.PRNGKey(10)
+        fake_dm_env.key = new_key
+        assert jnp.array_equal(new_key, fake_dm_env.key)
+    
+    def test_jumanji_environment_to_deep_mind_env__state_getter_setter(
+        self, fake_environment: FakeEnvironment, fake_state: FakeState
+        ) -> None:
+        """validate state getter and setter"""
```
Comment (clement-bonnet) [inline jumanji/wrappers_test.py:251]: Same here
```suggestion
        """Validates state getter and setters"""
```

File: jumanji/wrappers_test.py [hunk 6]
```diff
@@ -228,6 +235,25 @@ def test_jumanji_environment_to_deep_mind_env__unwrapped(
...
+        key = jax.random.PRNGKey(0)
+        fake_dm_env = JumanjiToDMEnvWrapper(fake_environment, key=key)
+        fake_dm_env.state = fake_state
+        assert fake_state == fake_dm_env.state
```
Comment (clement-bonnet) [inline jumanji/wrappers_test.py:255]: I think we even want to test reference equality here.
```suggestion
        assert fake_state is fake_dm_env.state
```

File: jumanji/wrappers_test.py [hunk 7]
```diff
@@ -21,6 +21,7 @@
 import jax
 import jax.numpy as jnp
 import jax.random as random
+from matplotlib.backend_bases import key_press_handler
```
Comment (clement-bonnet) [inline jumanji/wrappers_test.py:24]: Why is this needed?

File: jumanji/wrappers_test.py [hunk 8]
```diff
@@ -58,6 +59,12 @@ class MockWrapper(Wrapper[FakeState]):
     return MockWrapper
 
 
+@pytest.fixture
+def fake_state() -> FakeState:
+    return FakeState(key=jax.random.PRNGKey(0),
+                     step=jax.random.randint(jax.random.PRNGKey(0), [1], 0, 100))
+
+
```
Comment (clement-bonnet) [inline jumanji/wrappers_test.py:67]: If the fake state is used in only one test, it may be preferable not to make it a fixture and define it directly in the corresponding test. Also, I believe the step argument must rather be `step=jax.random.randint(jax.random.PRNGKey(0), (), 0, 100)`.

Comment (clement-bonnet) [reply jumanji/wrappers_test.py:67]: I would also change the key so that it's not the same as the other key defined in the `test_jumanji_environment_to_deep_mind_env__state_getter_setter` test.

**Review by clement-bonnet — CHANGES_REQUESTED** [review]
Thank you for your contribution and for the tests. Just left a few comments.

**Comment by clement-bonnet** [comment]
Thank you for your contribution! Looks good to me :rocket: 

It seems that you have not installed the pre-commit hooks, and hence the changes will probably not pass the linters (black at least). I can also see that tests are not executed on the pipeline for some reason (probably not your fault). Let me look into this!

**Review by clement-bonnet — APPROVED** [review]

File: jumanji/wrappers.py [hunk 9]
```diff
@@ -215,6 +215,22 @@ def action_spec(self) -> dm_env.specs.Array:
...
+    def state(self, new_state: State) -> None:
+        self._state = new_state
+
+    @property
+    def key(self) -> PRNGKey:
+        return self._key
+
+    @key.setter
+    def key(self, new_key: PRNGKey) -> None:
+        self._key = new_key
```
Comment (cemlyn007) [inline jumanji/wrappers.py:232]: Why are we using properties instead of attributes here?

**Review by cemlyn007 — CHANGES_REQUESTED** [review]

**Comment by clement-bonnet** [comment]
Is there anybody working on this? Otherwise, I'll be closing the PR for now.
